internal/extension/types/hooks.go

```diff
@@ -0,0 +1,35 @@
+package types
+
+import (
+	"github.com/envoyproxy/gateway/internal/ir"
+
+	"github.com/envoyproxy/gateway/internal/extension/proto"
+	xdsTypes "github.com/envoyproxy/gateway/internal/xds/types"
+)
+
+type HookType string
+
+const (
+	XDSHook          HookType = "xds"
+	InfraManagerHook HookType = "infra-manager"
+)
+
+// XDSHook specifies the entrypoints for extensions to hook into the XDS
+// translation pipeline.
+type XDSHookClient interface {
+	// PostHTTPListenerTranslation is the entrypoint for modifying XDS after processing
+	// the HTTP listener.
+	PostHTTPListenerTranslation(eCtx *proto.HTTPListenerExtensionContext, tCtx *xdsTypes.ResourceVersionTable) error
```

Comment:
Sort of a bikeshed: personally not a fan of these names because it requires additional mental load to figure out what they are referring to. `eCtx` makes a little bit of sense but the `tCtx` doesn't because its referring to untyped envoy config sources.

Second, Is the intent here to send the whole ResourceVersionTable for every listener? This would mean we would be serializing and sending over the wire a bunch of routes/listeners/clusters/ secrets that have nothing to do with it. Which made me just think...sending the ResourceVersionTable would also include secrets which I'm not so certain we want to be sending secrets over to an extension...

Reply:
tbh I'm not satisfied with the names either so if you have better names, I'm open to suggestions. `tCtx` is what is generally been used across the project to refer to the Envoy resource table

On the second point, remember that there are two interfaces `XDSHookClient` and the gRPC client interface to the extension service, call it `ExtensionXDSClient`.

`XDSHookClient` is the internal data structure that will be actually used by the translators, abtracting the gRPC transport - so its implementation will utilize `ExtensionXDSClient`. `ExtensionXDSClient` might have its own `PostHTTPListenerTranslation` method but with gRPC specific arguments

```go
type ExtensionXDSClient interface {
    PostHTTPListenerTranslation(ctx context.Context, in *PostHTTPListenerTranslationRequest, opts ...grpc.CallOption) (*PostHTTPListenerTranslationResponse, error)
}
```

`PostHTTPListenerTranslationRequest` would wrap `HTTPListenerExtensionContext`. Notice that we didn't send `ResourceVersionTable` over the wire. The extension would return `PostHTTPListenerTranslationResponse` that may outline some XDS-like data structure and the implementation of `XDSHookClient.PostHTTPListenerTranslation` would handle merging it with `ResourceVersionTable` after it receives the gRPC response.

Of course you can also do what you suggested and have the merge occur outside the method:
```go
tCtx := make(ResourceVersionTable)
// some stuff occurs
resp := xdsClient.PostHTTPListenerTranslation(eCtx)
// merge resp with tCtx
```

Reply:
update godoc